src/client/dbif.cc

```diff
@@ -0,0 +1,943 @@
+              child.second));
+        }
+
+        emit promise_iter->second->gotInfo(
+            QSharedPointer<dbif::ChildrenRequest::ReplyType>::create(objects));
+      } else {
+        if (iter_subscriptions == subscriptions_.end()) {
+          subchunk_data_requests_.erase(subchunkt_iter);
+        }
+
+        std::vector<data::ChunkDataItem> items;
+
+        for (auto child : *children_map) {
+          // FIXME
+          if(child.second.type() == dbif::ObjectType::CHUNK) {
+            items.push_back(data::ChunkDataItem::subchunk(0, 0, "",
+                QSharedPointer<NCObjectHandle>::create(child.second)));
+          } else if(child.second.type() == dbif::ObjectType::SUB_BLOB) {
+            items.push_back(data::ChunkDataItem::subblob("",
+                QSharedPointer<NCObjectHandle>::create(child.second)));
+          }
+        }
+
+        emit promise_iter->second->gotInfo(
+            QSharedPointer<dbif::ChunkDataRequest::ReplyType>::create(items));
+      }
+    }
+  }
+}
+
+void NCWrapper::handleRequestAckMessage(
+    std::shared_ptr<messages::MsgpackMsg> message) {
+  auto reply = std::dynamic_pointer_cast<messages::MsgRequestAck>(message);
+  if (reply) {
+    if (nc_->output() && detailed_debug_info_) {
+      *nc_->output() << QString("NCWrapper: received MsgRequestAck.") << endl;
+    }
+
+    const auto promise_iter = method_promises_.find(reply->rid);
+    if (promise_iter != method_promises_.end()) {
+      auto id_iter = created_objs_waiting_for_ack_.find(reply->rid);
+      if(id_iter != created_objs_waiting_for_ack_.end()) {
+        if (nc_->output() && detailed_debug_info_) {
+          *nc_->output() << QString("NCWrapper: file blob with id \"%1\""
+              " created.").arg(id_iter->second->id().toHexString()) << endl;
+        }
+        emit promise_iter->second->gotResult(
+            QSharedPointer<dbif::RootCreateFileBlobFromDataRequest::ReplyType>
+            ::create(id_iter->second));
+        created_objs_waiting_for_ack_.erase(id_iter);
+      }
+    }
+
+    // TODO Do we have other cases when we need to do something special when
+    // MsgRequestAck is received?
+  }
+}
+
+void getQStringAttr(std::shared_ptr<std::unordered_map<std::string,
+    std::shared_ptr<messages::MsgpackObject>>> attr, std::string key,
+    QString& val) {
+  auto iter = attr->find(key);
+  if (iter != attr->end()) {
+    std::shared_ptr<std::string> val_ptr;
+    messages::fromMsgpackObject(iter->second, val_ptr);
+    if (val_ptr) {
+      val = QString::fromStdString(*val_ptr);
+    }
+  }
+}
+
+template<class T> void getAttrSPtr(std::shared_ptr<std::unordered_map<
+    std::string, std::shared_ptr<messages::MsgpackObject>>> attr,
+    std::string key, T& val) {
+  auto iter = attr->find(key);
+  if (iter != attr->end()) {
+    std::shared_ptr<T> val_ptr;
+    messages::fromMsgpackObject(iter->second, val_ptr);
+    if (val_ptr) {
+      val = *val_ptr;
```

Comment:
We generally don't send pointers through network. We expect messages to contain simple values or something that is easy to handle/copy.